shell/platform/android/io/flutter/view/FlutterNativeView.java

```diff
@@ -149,116 +158,99 @@ public void setMessageHandler(String channel, BinaryMessageHandler handler) {
+                                mFlutterJNI.nativeInvokePlatformMessageResponseCallback(mNativePlatformView, replyId, reply, reply.position());
+                            }
                         }
-                        if (done.getAndSet(true)) {
-                            throw new IllegalStateException("Reply already submitted");
-                        }
-                        if (reply == null) {
-                            nativeInvokePlatformMessageEmptyResponseCallback(
-                                    mNativePlatformView, replyId);
-                        } else {
-                            nativeInvokePlatformMessageResponseCallback(
-                                    mNativePlatformView, replyId, reply, reply.position());
-                        }
-                    }
-                });
-            } catch (Exception ex) {
-                Log.e(TAG, "Uncaught exception in binary message listener", ex);
-                nativeInvokePlatformMessageEmptyResponseCallback(mNativePlatformView, replyId);
+                    });
+                } catch (Exception exception) {
+                    Log.e(TAG, "Uncaught exception in binary message listener", exception);
+                    mFlutterJNI.nativeInvokePlatformMessageEmptyResponseCallback(mNativePlatformView, replyId);
+                }
+                return;
             }
-            return;
+            mFlutterJNI.nativeInvokePlatformMessageEmptyResponseCallback(mNativePlatformView, replyId);
         }
-        nativeInvokePlatformMessageEmptyResponseCallback(mNativePlatformView, replyId);
-    }
 
-    // Called by native to respond to a platform message that we sent.
-    private void handlePlatformMessageResponse(int replyId, byte[] reply) {
-        BinaryReply callback = mPendingReplies.remove(replyId);
-        if (callback != null) {
-            try {
-                callback.reply(reply == null ? null : ByteBuffer.wrap(reply));
-            } catch (Exception ex) {
-                Log.e(TAG, "Uncaught exception in binary message reply handler", ex);
+        // Called by native to respond to a platform message that we sent.
+        public void handlePlatformMessageResponse(int replyId, byte[] reply) {
+            BinaryReply callback = mPendingReplies.remove(replyId);
+            if (callback != null) {
+                try {
+                    callback.reply(reply == null ? null : ByteBuffer.wrap(reply));
+                } catch (Exception ex) {
+                    Log.e(TAG, "Uncaught exception in binary message reply handler", ex);
+                }
             }
         }
     }
 
-    // Called by native to update the semantics/accessibility tree.
-    private void updateSemantics(ByteBuffer buffer, String[] strings) {
-        if (mFlutterView == null)
-            return;
-        mFlutterView.updateSemantics(buffer, strings);
-    }
+    private final class RenderSurfaceImpl implements RenderSurface {
```

Comment:
nit: might make this file a little more manageable if we move `RenderSurfaceImpl` to it's own file and giving it the dependencies it needs (just FlutterView?)

Reply:
This code was someone else's work so I just copied it, but my preference would be either to have FlutterNativeView implement RenderSurface directly, or declare an anonymous implementation of RenderSurface.  But I don't think extracting it would simplify anything because it's core to what FlutterNativeView does.

But should we be concerned about the simplicity of FlutterNativeView and FlutterView at this point?  Those classes will essentially be retired when the new embedding is used.  I don't expect that we'll be doing much work in these classes moving forward, and they're quite convoluted as it is.

Reply:
Can you give some more details about your eventual plan here? what will be implementing RenderSurface eventually?

Reply:
RenderSurface will be implemented by the new FlutterView.  It is essentially a delegate for behaviors that are required/expected of a Flutter rendering/ui surface.  You can see FlutterRenderer and RenderSurface in action in the staging branch:

https://github.com/flutter/engine/tree/21008_rewrite-android-embedding-to-reduce-coupling